board/rockchip/evb_rk3399/evb-rk3399.c

```diff
@@ -30,3 +38,75 @@ int board_early_init_f(void)
 	return 0;
 }
 #endif
+
+
+
+/* ALT_INFO Macro Function below added for DFU_ALT_INFO */
+
+#ifdef CONFIG_SET_DFU_ALT_INFO
+
+#define CONFIG_DFU_ALT_BOOT_SD \
+	"mmc 1=idbloader-img raw 0x40 0x2000;u-boot-itb raw 0x4000 0x2000\0"
+
+#define CONFIG_DFU_ALT_BOOT_EMMC \
+	"mmc 0=idbloader-img raw 0x40 0x2000;u-boot-itb raw 0x4000 0x2000\0"
+
+char *get_dfu_alt_boot(char *interface, char *devstr)
+{
+	char *alt_boot;
+	struct mmc *mmc;
+	int dev_num = 0;
+
+	if (devstr == NULL) {
+		devstr = "0";
+	}
+	alt_boot = env_get("dfu_alt_boot");
+
+	if (alt_boot == NULL) {
+		dev_num = simple_strtoul(devstr, NULL, 10);
+		mmc = find_mmc_device(dev_num);
+		if (!mmc) {
+			return NULL;
+		}
+		if (mmc_init(mmc)) {
+			return NULL;
+			}
+		if (IS_SD(mmc)) {
+			alt_boot = CONFIG_DFU_ALT_BOOT_SD;
+		} else {
+			alt_boot = CONFIG_DFU_ALT_BOOT_EMMC;
+		}
+	}
+	return alt_boot;
+}
+#endif
+
+#ifdef CONFIG_SET_DFU_ALT_INFO
+#define CONFIG_SET_DFU_ALT_BUF_LEN (SZ_1K)
+void set_dfu_alt_info(char *interface, char *devstr)
+{
+	size_t buf_size = CONFIG_SET_DFU_ALT_BUF_LEN;
+
+	ALLOC_CACHE_ALIGN_BUFFER(char, buf, buf_size);
+	char *alt_info = "Settings not found!";
+	char *status = "error!\n";
+	char *alt_setting;
```

Comment:
It looks like you copied this code from the Samsung implementation, but it isn't a good example. This code should not be setting the dfu_alt_info value to "Settings not found!". If the alt info cannot be set, then it should be left unset. Much of this code can be cleaned up.

This can be made a lot simpler. All that is needed is to choose between two different values for dfu_alt_info depending on whether you're booting from SD or eMMC. So, drop the `dfu_alt_boot` variable and just env_set one of the two strings returned by get_dfu_alt_boot(). In fact, all of this should only be one short function.

Reply:
set_dfu_alt_info
The function changes are required only for reflecting hardcoded setting and may not be valid case for user.
In user scenario the hard coded dfu_alt partitions may point to different user parttion and may lead to corruption of the partiton.
This part of the code is not required for actual scenario.
So we suggest user to manually set the env variable "dfu_alt_info" as per the partition table.  
Can we remove this option itself and keep only manual setting option for capsule update. 
Kindly suggest


Modified function:
/* ALT_INFO Macro Function below added for DFU_ALT_INFO */
#ifdef CONFIG_SET_DFU_ALT_INFO
#define CONFIG_DFU_ALT_BOOT_SD \
	"mmc 1=idbloader-img raw 0x40 0x2000;u-boot-itb raw 0x4000 0x2000\0"
#define CONFIG_DFU_ALT_BOOT_EMMC \
	"mmc 0=idbloader-img raw 0x40 0x2000;u-boot-itb raw 0x4000 0x2000\0"
void set_dfu_alt_info(char *interface, char *devstr)
{
	char *alt_info;
	struct mmc *mmc;
	int dev_num = 0;
	alt_info = env_get("dfu_alt_info");
	if (!devstr) 
		devstr = "0";
	if (!alt_info) {
		dev_num = simple_strtoul(devstr, NULL, 10);
		mmc = find_mmc_device(dev_num);
		if (!mmc || mmc_init(mmc)) 
			return;
		alt_info = IS_SD(mmc) ? CONFIG_DFU_ALT_BOOT_SD : CONFIG_DFU_ALT_BOOT_EMMC;
		}
	env_set("dfu_alt_info", alt_info);
}
#endif